src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -424,29 +442,317 @@ Shard::valFetch(uint256 const& hash)
+                boost::str(boost::format("PRAGMA cache_size=-%d;") %
+                kilobytes(config.getSize(siLgrDBCache)));
+            lgrSQLiteDB_->setupCheckpointing(&app_.getJobQueue(), app_.logs());
+
+            txSQLiteDB_ = std::make_unique <DatabaseCon>(
+                setup,
+                TxDBName,
+                TxDBPragma,
+                TxDBInit);
+            txSQLiteDB_->getSession() <<
+                boost::str(boost::format("PRAGMA cache_size=-%d;") %
+                kilobytes(config.getSize(siTxnDBCache)));
+            txSQLiteDB_->setupCheckpointing(&app_.getJobQueue(), app_.logs());
         }
     }
     catch (std::exception const& e)
     {
         JLOG(j_.error()) <<
-            "exception: " << e.what();
+            "shard " << index_ <<
+            " exception " << e.what() <<
+            " in function " << __func__;
+        return false;
     }
-    return nObj;
+    return true;
+}
+
+bool
+Shard::setSQLiteStored(std::shared_ptr<Ledger const> const& ledger)
+{
+    auto const seq {ledger->info().seq};
+    assert(backend_ && !complete_);
+    assert(!boost::icl::contains(storedSeqs_, seq));
+
+    {
+        auto& session {txSQLiteDB_->getSession()};
+        soci::transaction tr(session);
+
+        session <<
+            "DELETE FROM Transactions WHERE LedgerSeq = :seq;"
+            , soci::use(seq);
+        session <<
+            "DELETE FROM AccountTransactions WHERE LedgerSeq = :seq;"
+            , soci::use(seq);
+
+        if (ledger->info().txHash.isNonZero())
+        {
+            auto const sSeq {std::to_string(seq)};
+            if (!ledger->txMap().isValid())
+            {
+                JLOG(j_.error()) <<
+                    "shard " << index_ <<
+                    " has an invalid transaction map" <<
+                    " on sequence " << sSeq;
+                return false;
+            }
+
+            for (auto const& item : ledger->txs)
+            {
+                auto const txID {item.first->getTransactionID()};
+                auto const sTxID {to_string(txID)};
+                auto txMeta {std::make_shared<TxMeta>(
+                    txID, ledger->seq(), *item.second)};
+
+                session <<
+                    "DELETE FROM AccountTransactions WHERE TransID = :txID;"
+                    , soci::use(sTxID);
+
+                auto const& accounts = txMeta->getAffectedAccounts(j_);
+                if (!accounts.empty())
+                {
+                    auto const sTxSeq {std::to_string(txMeta->getIndex())};
+                    std::string sql (
+                        "INSERT INTO AccountTransactions "
+                        "(TransID, Account, LedgerSeq, TxnSeq) VALUES ('");
+
+                    // Try to make an educated guess on how much space we'll
+                    // need for our arguments. In argument order we have:
+                    // 64 + 34 + 10 + 10 = 118 + 10 extra = 128 bytes
```

Comment:
possible alternative for your consideration:

```
std::string sql (
    "INSERT INTO AccountTransactions "
    "(TransID, Account, LedgerSeq, TxnSeq) VALUES ");
sql.reserve(sql.length() + (accounts.size() * 128));
boost::algorithm::join (accounts |
    boost::adaptors::transformed(
        [&](AccountID const& a){
            return boost::str (boost::format (
                "('%1%','%2%',%3%,%4%)")
                % sTxID
                % ripple::toBase58(a)
                % sSeq
                % sTxSeq); }),
    ",");
sql += ';';
session << sql;
```

Reply:
Nice use of join and transformed, thanks!